src/livekit/MediaDevicesContext.tsx

```diff
@@ -249,6 +253,144 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
     [setNumCallersUsingNames],
   );
 
+  return {
+    audioInput,
+    videoInput,
+    startUsingDeviceNames,
+    stopUsingDeviceNames,
+    usingNames,
+  };
+}
+
+interface Props {
+  children: JSX.Element;
+}
+
+export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+  const {
+    audioInput,
+    videoInput,
+    startUsingDeviceNames,
+    stopUsingDeviceNames,
+    usingNames,
+  } = useInputDevices();
+
+  const audioOutput = useMediaDevice(
+    "audiooutput",
+    audioOutputSetting,
+    usingNames,
+  );
+
+  const context: MediaDevices = useMemo(
+    () => ({
+      audioInput,
+      audioOutput,
+      videoInput,
+      startUsingDeviceNames,
+      stopUsingDeviceNames,
+    }),
+    [
+      audioInput,
+      audioOutput,
+      videoInput,
+      startUsingDeviceNames,
+      stopUsingDeviceNames,
+    ],
+  );
+
+  return (
+    <MediaDevicesContext.Provider value={context}>
+      {children}
+    </MediaDevicesContext.Provider>
+  );
+};
+
+function useControlledOutput(): MediaDevice {
+  const showEarpiece = useShowEarpiece();
+
+  const available = useObservableEagerState(
+    useObservable(() =>
+      setOutputDevices$.pipe(
+        startWith<OutputDevice[]>([]),
+        map((devices) => {
+          const devicesMap = new Map<string, DeviceLabel>(
+            devices.map(({ id, name }) => [id, { type: "name", name }]),
+          );
+          if (showEarpiece)
+            devicesMap.set(EARPIECE_CONFIG_ID, { type: "earpiece" });
+          return devicesMap;
+        }),
+      ),
+    ),
```

Comment:
It's not safe to reference `showEarpiece` here inside `useObservable` due to https://observable-hooks.js.org/guide/gotchas.html#epic-closure

Instead you could pass `showEarpiece` to the inputs array of `useObservable` like so:

```suggestion
    useObservable((showEarpiece$) =>
      combineLatest([setOutputDevices$.pipe(startWith<OutputDevice[]>([])), showEarpiece$], (devices, showEarpiece) => {
        const devicesMap = new Map<string, DeviceLabel>(
          devices.map(({ id, name }) => [id, { type: "name", name }]),
        );
        if (showEarpiece)
          devicesMap.set(EARPIECE_CONFIG_ID, { type: "earpiece" });
        return devicesMap;
      }),
      [showEarpiece],
    ),
```

Or you could even convert `useShowEarpiece` from a React hook into an Observable, if that feels nicer.

Oh, or I suppose replacing `useObservable` with `useMemo` would work too. Lots of options :)

Reply:
I ended up witht he combine latest approach